### PR TITLE
[workflow] Get rid of asyncio.run for py36

### LINKED ISSUE #17064: [Workflow] Potential issues in workflow
### What is the problem?

asyncio shows some warnings I did not expect (w/ python 3.7.10):

```
(pid=87344) /Users/s.z/Code/ray/python/ray/worker.py:421: RuntimeWarning: coroutine 'WorkflowStorage._update_dynamic_output' was never awaited
(pid=87344) RuntimeWarning: Enable tracemalloc to get the object allocation traceback
(pid=87344) /Users/s.z/Code/ray/python/ray/worker.py:421: RuntimeWarning: coroutine 'FilesystemStorageImpl.save_step_output' was never awaited
```

### Reproduction (REQUIRED)

It can be reproduced by running workflow tests

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

Py36 doesn't have asyncio.run. We should avoid using this one until we end the support for py36.

Besides this, also try to run recover tests which is blocked by a bug in workflow. That part will be added later.
## Related issue number

Closes #17064
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
It looks like CI doesn't trigger the actual path of s3 for test_workflow_storage. I'll investigate it.

File: python/ray/experimental/workflow/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from moto import mock_s3
+from mock_server import *  # noqa
+from ray.experimental.workflow import storage
+
+
+@pytest.fixture(scope="function")
+def filesystem_storage():
+    # TODO: use tmp path once fixed the path issues
+    storage.set_global_storage(
+        storage.create_storage("/tmp/ray/workflow_data/"))
```
Comment (iycheng) [inline python/ray/experimental/workflow/tests/conftest.py:12]: @suquark there is a bug here about storage.
Basically, if the pash is not /tmp/ray/workflow_data, the recover test will fail.
I wanted to add tests of s3 for recover test for a better coverage (https://gist.github.com/iycheng/aae32fe2c4f73c71a4f2eec357549967), but with s3 it still use the file system based global storage.


Could you help check this issue? Basically, specify the storage explicitly with different paths should work.

After that, I'll add tests for recover path. But this PR is ready for ready which can unblock you.

File: python/ray/experimental/workflow/storage/base.py [hunk 2]
```diff
@@ -39,8 +39,8 @@ class Storage(metaclass=abc.ABCMeta):
     """
 
     @abstractmethod
-    async def load_step_input_metadata(self, workflow_id: str,
-                                       step_id: StepID) -> Dict[str, Any]:
+    async def load_step_input_metadata(self, workflow_id: str, step_id: StepID
+                                       ) -> Awaitable[Dict[str, Any]]:
```
Comment (suquark) [inline python/ray/experimental/workflow/storage/base.py:43]: sorry, but there may be some misconception: we should not use awaitable for async function return type. here are some official examples: https://github.com/python/cpython/pull/20386/files

**Comment by suquark** [comment]
there is no `await` at the end of `_write_step_inputs`. could you fix this and check if this address your issue?

**Comment by iycheng** [comment]
> there is no `await` at the end of `_write_step_inputs`, this could fail the s3 recovery test (it did not affect fs because it is blocking). could you fix this and check if this address your issue?
> 
> btw, if you see warnings like in #17064, something is going wrong. there is also a debug switch in asyncio if I remember correctly

Right now I can't run the recovery code path with s3 due to the bug in workflow. I'll add tests once it got fixed.

**Review by suquark — APPROVED** [review]
LGTM

**Comment by suquark** [comment]
@iycheng you can merge this PR first (or I can do it for you). I created an issue about recovery issues. I can handle it later.

**Pull request merged**